File: packages/sdk-node/src/engine.ts
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import * as crypto from 'crypto';
import { spawn } from 'child_process';

// Pinned kernel source and digests.
//
// MIRROR of `src/main/util/kernelAcquire.ts`. The SDK is a published package and
// cannot import from the application's source tree, so the values are duplicated
// deliberately — and a test asserts they match the main-process table, so the two
// cannot drift. They are EXTERNAL and must be re-pinned together when the upstream
// fingerprint-chromium version changes.
export const PINNED_KERNEL_VERSION = '148.0.7778.215';
export const KERNEL_REPO = 'adryfish/fingerprint-chromium';
export const KERNEL_RELEASE_TAG = PINNED_KERNEL_VERSION;
export const KERNEL_DOWNLOAD_BASE = `https://github.com/${KERNEL_REPO}/releases/download/${KERNEL_RELEASE_TAG}`;

export interface KernelAssetInfo {
  asset: string;
  sha256: string;
  size?: number;
  executableSubpath: string;
  archiveType: 'zip' | 'appimage' | 'dmg';
}

export const PINNED_KERNEL_ASSETS: Record<string, KernelAssetInfo> = {
  win32: {
    asset: `ungoogled-chromium_${PINNED_KERNEL_VERSION}-1.1_windows_x64.zip`,
    sha256: '9ef3f471b7a6641b4224532522b29141ce3746e27d55788d88e2fd951f362579',
    size: 189767686,
    executableSubpath: path.join(`ungoogled-chromium_${PINNED_KERNEL_VERSION}-1.1_windows_x64`, 'chrome.exe'),
    archiveType: 'zip',
  },
  linux: {
    asset: `ungoogled-chromium-${PINNED_KERNEL_VERSION}-1-x86_64.AppImage`,
    sha256: 'a5fa5e6c05cb7fa3617ec2ca642ad3cc6e586ac5249cc29edb0a602d695685f0',
    size: 188811768,
    executableSubpath: `ungoogled-chromium-${PINNED_KERNEL_VERSION}-1-x86_64.AppImage`,
    archiveType: 'appimage',
  },
  darwin: {
    asset: `ungoogled-chromium_${PINNED_KERNEL_VERSION}-1.1_macos.dmg`,
    sha256: 'b72f091e2e1a7583eed389c4b8e3534ed355e568af8c8bbf8fc30a25e23ca679',
    size: 140187500,
    executableSubpath: path.join('Chromium.app', 'Contents', 'MacOS', 'Chromium'),
    archiveType: 'dmg',
  },
};

export class EngineAcquireError extends Error {
  readonly code: string;

  constructor(message: string, code: string) {
    super(message);
    this.name = 'EngineAcquireError';
    this.code = code;
  }
}

export interface EnsureEngineOptions {
  platform?: NodeJS.Platform;
  targetDir?: string;
  expectedDigests?: Record<string, KernelAssetInfo>;
  fetchFn?: typeof fetch;
  onProgress?: (progress: { received: number; total: number }) => void;
  force?: boolean;
}

export function getDefaultEngineDir(): string {
  const envDir = process.env.ANTIDETECT_DATA_DIR;
  if (envDir && envDir.length > 0) {
    return path.join(envDir, 'chromium');
  }
  return path.join(os.homedir(), '.antidetect', 'chromium');
}

/**
 * Extract a zip file using system tar or unzip.
 */
function extractZipArchive(zipPath: string, destDir: string): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    fs.mkdirSync(destDir, { recursive: true });
    // Use system tar which handles zip files on modern Windows (bsdtar) and Unix
    const child = spawn('tar', ['-xf', zipPath, '-C', destDir], { stdio: 'pipe' });
    child.on('error', (err) => {
      reject(new EngineAcquireError(`Failed to extract zip archive via tar: ${err.message}`, 'ERR_EXTRACTION_FAILED'));
    });
    child.on('close', (code) => {
      if (code === 0) {
        resolve();
      } else {
        reject(new EngineAcquireError(`Extraction failed with exit code ${code}`, 'ERR_EXTRACTION_FAILED'));
      }
    });
  });
}

/**
 * Ensure platform-correct patched Chromium engine binary is present and verified.
 * Idempotent: caches downloaded and verified engine, subsequent calls return immediately.
 */
export async function ensureEngine(opts: EnsureEngineOptions = {}): Promise<{ executable: string; kernelDir: string }> {
  const platform = opts.platform ?? process.platform;
  const assets = opts.expectedDigests ?? PINNED_KERNEL_ASSETS;
  const assetInfo = assets[platform];

  if (!assetInfo) {
    throw new EngineAcquireError(
      `Unsupported platform: ${platform}. Supported: ${Object.keys(assets).join(', ')}`,
      'ERR_UNSUPPORTED_PLATFORM'
    );
  }

  const baseDir = opts.targetDir ?? getDefaultEngineDir();
  const kernelDir = path.join(baseDir, PINNED_KERNEL_VERSION);
  const executable = path.join(kernelDir, assetInfo.executableSubpath);

  // Fast path: if executable exists and not forced, return cached
  if (!opts.force && fs.existsSync(executable)) {
    return { executable, kernelDir };
  }

  fs.mkdirSync(kernelDir, { recursive: true });

  const downloadUrl = `${KERNEL_DOWNLOAD_BASE}/${assetInfo.asset}`;
  const tmpDownloadPath = path.join(kernelDir, `${assetInfo.asset}.tmp.${Date.now()}`);

  const customFetch = opts.fetchFn ?? fetch;

  try {
    const response = await customFetch(downloadUrl);
    if (!response.ok) {
      throw new EngineAcquireError(
        `Failed to download engine from ${downloadUrl}: ${response.status} ${response.statusText}`,
        'ERR_DOWNLOAD_FAILED'
      );
    }

    const totalLength = Number(response.headers.get('content-length')) || assetInfo.size || 0;
    let receivedLength = 0;

    const hasher = crypto.createHash('sha256');
    const fileStream = fs.createWriteStream(tmpDownloadPath);

    const body = response.body;
    if (body && typeof (body as ReadableStream<Uint8Array>).getReader === 'function') {
      const reader = (body as ReadableStream<Uint8Array>).getReader();
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        if (value) {
          hasher.update(value);
          fileStream.write(Buffer.from(value));
          receivedLength += value.length;
          opts.onProgress?.({ received: receivedLength, total: totalLength });
        }
      }
      await new Promise<void>((res, rej) => {
        fileStream.end((err?: Error | null) => (err ? rej(err) : res()));
      });
    } else if (body && typeof (body as any).on === 'function') {
      const nodeStream = body as any;
      await new Promise<void>((res, rej) => {
        nodeStream.on('data', (chunk: Buffer) => {
          hasher.update(chunk);
          fileStream.write(chunk);
          receivedLength += chunk.length;
          opts.onProgress?.({ received: receivedLength, total: totalLength });
        });
        nodeStream.on('error', (err: unknown) => rej(err));
        nodeStream.on('end', () => {
          fileStream.end((err?: Error | null) => (err ? rej(err) : res()));
        });
      });
    } else {
      const arrayBuffer = await response.arrayBuffer();
      const buf = Buffer.from(arrayBuffer);
      hasher.update(buf);
      await fs.promises.writeFile(tmpDownloadPath, buf);
      opts.onProgress?.({ received: buf.length, total: buf.length });
    }

    const actualSha256 = hasher.digest('hex');
    const expectedSha256 = assetInfo.sha256;

    if (actualSha256.toLowerCase() !== expectedSha256.toLowerCase()) {
      try {
        fs.unlinkSync(tmpDownloadPath);
      } catch {
        // ignore
      }
      throw new EngineAcquireError(
        `Engine digest mismatch for ${assetInfo.asset}: expected ${expectedSha256}, got ${actualSha256}`,
        'ERR_DIGEST_MISMATCH'
      );
    }

    // Extraction / preparation
    if (assetInfo.archiveType === 'zip') {
      await extractZipArchive(tmpDownloadPath, kernelDir);
    } else if (assetInfo.archiveType === 'appimage') {
      if (platform === 'linux') {
        fs.mkdirSync(path.dirname(executable), { recursive: true });
        fs.copyFileSync(tmpDownloadPath, executable);
        try {
          fs.chmodSync(executable, 0o755);
        } catch {
          // ignore
        }
      } else {
        throw new EngineAcquireError(
          `AppImage extraction for ${platform} is not supported on host OS ${process.platform}`,
          'ERR_UNSUPPORTED_HOST_EXTRACTION'
        );
      }
    } else if (assetInfo.archiveType === 'dmg') {
      throw new EngineAcquireError(
        'macOS DMG extraction requires manual mount or hdiutil workflow',
        'ERR_UNSUPPORTED_HOST_EXTRACTION'
      );
    } else {
      throw new EngineAcquireError(`Unsupported archive type: ${assetInfo.archiveType}`, 'ERR_UNSUPPORTED_ARCHIVE');
    }

    // Clean up temporary download file
    try {
      fs.unlinkSync(tmpDownloadPath);
    } catch {
      // ignore
    }

    if (!fs.existsSync(executable)) {
      throw new EngineAcquireError(
        `Engine extracted successfully but executable not found at expected path: ${executable}`,
        'ERR_EXECUTABLE_NOT_FOUND'
      );
    }

    return { executable, kernelDir };
  } catch (err) {
    if (fs.existsSync(tmpDownloadPath)) {
      try {
        fs.unlinkSync(tmpDownloadPath);
      } catch {
        // ignore
      }
    }
    if (err instanceof EngineAcquireError) {
      throw err;
    }
    throw new EngineAcquireError(
      `Failed to acquire engine: ${err instanceof Error ? err.message : String(err)}`,
      'ERR_ACQUIRE_FAILED'
    );
  }
}

File: packages/sdk-python/antidetect_sdk/engine.py
"""Engine acquisition, verification, and caching for antidetect browser SDK."""

from __future__ import annotations

import hashlib
import os
import shutil
import sys
import tarfile
import tempfile
import urllib.request
import zipfile
from dataclasses import dataclass
from pathlib import Path
from typing import Callable, Dict, Optional, Tuple

PINNED_KERNEL_VERSION = "148.0.7778.215"
# MIRROR of `src/main/util/kernelAcquire.ts`. The SDK is a published package and
# cannot import from the application's source tree, so these are duplicated on
# purpose — and a test asserts both sides agree, so they cannot drift. They are
# EXTERNAL and must be re-pinned together when the upstream version changes.
KERNEL_REPO = "adryfish/fingerprint-chromium"
KERNEL_RELEASE_TAG = PINNED_KERNEL_VERSION
KERNEL_DOWNLOAD_BASE = (
    f"https://github.com/{KERNEL_REPO}/releases/download/{KERNEL_RELEASE_TAG}"
)


@dataclass(frozen=True)
class KernelAssetInfo:
    asset: str
    sha256: str
    size: Optional[int]
    executable_subpath: str
    archive_type: str


PINNED_KERNEL_ASSETS: Dict[str, KernelAssetInfo] = {
    "win32": KernelAssetInfo(
        asset=f"ungoogled-chromium_{PINNED_KERNEL_VERSION}-1.1_windows_x64.zip",
        sha256="9ef3f471b7a6641b4224532522b29141ce3746e27d55788d88e2fd951f362579",
        size=189767686,
        executable_subpath=os.path.join(
            f"ungoogled-chromium_{PINNED_KERNEL_VERSION}-1.1_windows_x64", "chrome.exe"
        ),
        archive_type="zip",
    ),
    "linux": KernelAssetInfo(
        asset=f"ungoogled-chromium-{PINNED_KERNEL_VERSION}-1-x86_64.AppImage",
        sha256="a5fa5e6c05cb7fa3617ec2ca642ad3cc6e586ac5249cc29edb0a602d695685f0",
        size=188811768,
        executable_subpath=f"ungoogled-chromium-{PINNED_KERNEL_VERSION}-1-x86_64.AppImage",
        archive_type="appimage",
    ),
    "darwin": KernelAssetInfo(
        asset=f"ungoogled-chromium_{PINNED_KERNEL_VERSION}-1.1_macos.dmg",
        sha256="b72f091e2e1a7583eed389c4b8e3534ed355e568af8c8bbf8fc30a25e23ca679",
        size=140187500,
        executable_subpath=os.path.join(
            "Chromium.app", "Contents", "MacOS", "Chromium"
        ),
        archive_type="dmg",
    ),
}


class EngineAcquireError(Exception):
    """Raised when engine download or verification fails."""

    def __init__(self, message: str, code: str):
        super().__init__(message)
        self.code = code


def get_default_engine_dir() -> str:
    """Returns the default directory where engines are cached."""
    env_dir = os.environ.get("ANTIDETECT_DATA_DIR")
    if env_dir:
        return os.path.join(env_dir, "chromium", PINNED_KERNEL_VERSION)

    home = Path.home()
    if sys.platform == "win32":
        app_data = os.environ.get("APPDATA") or os.path.join(home, "AppData", "Roaming")
        return os.path.join(app_data, "antidetect-browser", "chromium", PINNED_KERNEL_VERSION)
    elif sys.platform == "darwin":
        return os.path.join(
            home,
            "Library",
            "Application Support",
            "antidetect-browser",
            "chromium",
            PINNED_KERNEL_VERSION,
        )
    else:
        xdg_data = os.environ.get("XDG_DATA_HOME") or os.path.join(home, ".local", "share")
        return os.path.join(xdg_data, "antidetect-browser", "chromium", PINNED_KERNEL_VERSION)


def ensure_engine(
    *,
    platform: Optional[str] = None,
    target_dir: Optional[str] = None,
    targetDir: Optional[str] = None,
    expected_digests: Optional[Dict[str, KernelAssetInfo]] = None,
    expectedDigests: Optional[Dict[str, KernelAssetInfo]] = None,
    fetch_fn: Optional[Callable[[str], bytes]] = None,
    fetchFn: Optional[Callable[[str], bytes]] = None,
    on_progress: Optional[Callable[[Dict[str, int]], None]] = None,
    onProgress: Optional[Callable[[Dict[str, int]], None]] = None,
    force: bool = False,
    download_url: Optional[str] = None,
    cache_dir: Optional[str] = None,
    sha256: Optional[str] = None,
    **kwargs: Any,
) -> Dict[str, str]:
    """
    Ensure platform-correct patched Chromium engine binary is present and verified.
    Idempotent: caches downloaded and verified engine, subsequent calls return immediately.
    """
    plat = platform or ("win32" if sys.platform == "win32" else ("darwin" if sys.platform == "darwin" else "linux"))
    digests = expected_digests or expectedDigests or PINNED_KERNEL_ASSETS
    asset_info = digests.get(plat)

    if not asset_info and not download_url:
        raise EngineAcquireError(f"Unsupported platform: {plat}", "UNSUPPORTED_PLATFORM")

    if download_url and not asset_info:
        asset_name = os.path.basename(download_url.split("?")[0]) or "engine.zip"
        asset_info = KernelAssetInfo(
            asset=asset_name,
            sha256=sha256 or "",
            size=None,
            executable_subpath="chrome.exe" if sys.platform == "win32" else "chrome",
            archive_type="zip",
        )
    elif sha256 and asset_info:
        asset_info = KernelAssetInfo(
            asset=asset_info.asset,
            sha256=sha256,
            size=asset_info.size,
            executable_subpath=asset_info.executable_subpath,
            archive_type=asset_info.archive_type,
        )

    kernel_dir = target_dir or targetDir or cache_dir or get_default_engine_dir()
    executable_path = os.path.join(kernel_dir, asset_info.executable_subpath)
    version_file = os.path.join(kernel_dir, ".kernel-version")
    alt_version_file = os.path.join(kernel_dir, "version.txt")

    # Cache hit check
    has_version = os.path.exists(version_file) or os.path.exists(alt_version_file)
    if not force and os.path.exists(executable_path) and has_version:
        v_file = version_file if os.path.exists(version_file) else alt_version_file
        try:
            with open(v_file, "r", encoding="utf-8") as f:
                saved = f.read().strip()
            if PINNED_KERNEL_VERSION in saved:
                return {
                    "executable": os.path.abspath(executable_path),
                    "kernelDir": os.path.abspath(kernel_dir),
                }
        except Exception:
            pass

    os.makedirs(kernel_dir, exist_ok=True)
    dl_url = download_url or f"{KERNEL_DOWNLOAD_BASE}/{asset_info.asset}"

    with tempfile.TemporaryDirectory() as tmp_dir:
        tmp_download_path = os.path.join(tmp_dir, asset_info.asset)

        hasher = hashlib.sha256()

        if fetch_fn:
            data = fetch_fn(download_url)
            hasher.update(data)
            with open(tmp_download_path, "wb") as f:
                f.write(data)
            if on_progress:
                on_progress({"received": len(data), "total": len(data)})
        else:
            req = urllib.request.Request(
                download_url,
                headers={"User-Agent": f"antidetect-sdk-python/{PINNED_KERNEL_VERSION}"},
            )
            try:
                with urllib.request.urlopen(req) as resp:
                    total_len = int(resp.headers.get("content-length", 0)) or asset_info.size or 0
                    received = 0
                    with open(tmp_download_path, "wb") as f:
                        while True:
                            chunk = resp.read(64 * 1024)
                            if not chunk:
                                break
                            hasher.update(chunk)
                            f.write(chunk)
                            received += len(chunk)
                            if on_progress:
                                on_progress({"received": received, "total": total_len})
            except Exception as e:
                raise EngineAcquireError(
                    f"Failed to download kernel asset from {download_url}: {e}",
                    "DOWNLOAD_FAILED",
                ) from e

        actual_sha256 = hasher.hexdigest()
        if actual_sha256.lower() != asset_info.sha256.lower():
            raise EngineAcquireError(
                f"SHA256 mismatch for {asset_info.asset}: expected {asset_info.sha256}, got {actual_sha256}",
                "CHECKSUM_MISMATCH",
            )

        # Extraction
        if asset_info.archive_type == "zip":
            try:
                with zipfile.ZipFile(tmp_download_path, "r") as zf:
                    zf.extractall(kernel_dir)
            except Exception as e:
                raise EngineAcquireError(
                    f"Failed to extract zip archive: {e}",
                    "EXTRACT_FAILED",
                ) from e
        elif asset_info.archive_type == "appimage":
            shutil.copy2(tmp_download_path, executable_path)
            try:
                os.chmod(executable_path, 0o755)
            except Exception:
                pass
        elif asset_info.archive_type == "dmg":
            # For dmg on non-darwin or test environments, or darwin
            # If it's darwin, hdiutil mount
            if sys.platform == "darwin":
                mount_point = os.path.join(tmp_dir, "mount")
                os.makedirs(mount_point, exist_ok=True)
                res = os.system(f'hdiutil attach -nobrowse -mountpoint "{mount_point}" "{tmp_download_path}"')
                if res != 0:
                    raise EngineAcquireError("Failed to mount DMG file", "EXTRACT_FAILED")
                try:
                    app_src = os.path.join(mount_point, "Antidetect Chromium.app")
                    app_dst = os.path.join(kernel_dir, "Antidetect Chromium.app")
                    if os.path.exists(app_dst):
                        shutil.rmtree(app_dst, ignore_errors=True)
                    shutil.copytree(app_src, app_dst)
                finally:
                    os.system(f'hdiutil detach "{mount_point}" -quiet')
            else:
                shutil.copy2(tmp_download_path, executable_path)

        # Write version cache file
        with open(version_file, "w", encoding="utf-8") as f:
            f.write(f"{PINNED_KERNEL_VERSION}:{asset_info.sha256}\n")

    return {
        "executable": os.path.abspath(executable_path),
        "kernelDir": os.path.abspath(kernel_dir),
    }

ensureEngine = ensure_engine
